### PR TITLE
Deduplicate diagnostic publishing prologue/epilogue behind a canonical helper (#510)

### LINKED ISSUE #510: Deduplicate diagnostic publishing prologue/epilogue behind a canonical helper
## Summary

`publish_feature_diagnostics` and `publish_rust_diagnostics` share a near-identical
prologue (client guard → `Url::from_file_path` guard) and epilogue
(`PublishDiagnosticsParams::new` → `client.notify` → error log). They differ only
in the index lookup and the set of `compute_*` calls. ~35 lines of duplicated
scaffolding around the publish boundary.

## Locations

| Function | File | Line |
|---|---|---|
| `publish_feature_diagnostics` | `crates/rstest-bdd-server/src/handlers/diagnostics/publish.rs` | 28 |
| `publish_rust_diagnostics` | `crates/rstest-bdd-server/src/handlers/diagnostics/publish.rs` | 81 |

## Canonical-implementation mandate

- [ ] Extract one canonical helper holding the client/URI guards and the
      notify/error-log epilogue, e.g.
      `fn publish_with(state: &ServerState, path: &Path, compute: impl FnOnce(&ServerState, &Path) -> Vec<Diagnostic>)`.
      The feature variant's `feature_index` guard is supplied via the closure (or
      a thin wrapper).
- [ ] Rewrite both public functions to delegate to the canonical helper; the
      publish boundary (guards, params construction, notify) exists once.
- [ ] Document the canonical publish path in `docs/developers-guide.md`
      (ownership, permitted call-sites), per the abstraction policy in
      `AGENTS.md`.

## Testing requirements

- [ ] **Snapshot assertions** (`insta`): snapshot the published
      `PublishDiagnosticsParams` (URI + ordered diagnostics) for a representative
      feature file and a representative Rust file, so the canonical path's output
      is pinned across the refactor.
- [ ] **Property tests** (`proptest`): assert the publish invariants hold for
      arbitrary diagnostic vectors — an empty vector still publishes (clearing
      prior diagnostics), and the published count equals the computed count.
- [ ] **Model checking**: not applicable.

## Acceptance criteria

- [ ] One canonical publish helper; both functions delegate to it.
- [ ] Developer's guide updated.
- [ ] Snapshot and property tests added and passing.
- [ ] `make check-fmt`, `make lint`, `make test` pass.

## Context

Raised from a workspace refactoring audit. Tier 2 finding. Requested by @leynos.

### PR DESCRIPTION
**Pull request by leynos**
## Summary

Closes #510

- Extract one canonical `publish_with(state, path, failure_message, compute)` helper owning the client guard, URI guard, `PublishDiagnosticsParams` construction, `client.notify`, and error logging; `publish_feature_diagnostics` and `publish_rust_diagnostics` are now thin wrappers over it.
- The feature variant's `feature_index` guard is supplied via its compute closure: `None` skips publishing entirely (preserving previous diagnostics), while `Some(vec![])` still publishes so stale diagnostics are cleared.
- `prepare_publish` separates the pure parameter construction from the notify side effect for testability without a client socket.
- Tests:
  - `insta` snapshots pin the published `PublishDiagnosticsParams` (URI + ordered diagnostics) for a representative feature file and Rust file (temp dir normalized to `[TMP]`).
  - A `proptest` suite pins the publish invariants for arbitrary diagnostic vectors: count and order preserved, empty vector still publishes, URI targets the requested path.
- Canonical publish path documented in `docs/developers-guide.md` (ownership, permitted call-sites, composition rules).

## Testing

- `make check-fmt`, `make lint`, `make typecheck`, `make test` (1490 passed), `make markdownlint` all pass.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by Sourcery

Unify LSP diagnostic publishing behind a shared helper and add tests and docs around the canonical publish path.

Enhancements:
- Introduce shared `publish_with` and `prepare_publish` helpers to centralize client/URI guarding, parameter construction, and error logging for diagnostic publishing.
- Split feature and Rust diagnostic computation into dedicated functions to clarify behavior when no feature index exists and to always clear stale Rust diagnostics.

Build:
- Add `insta` with filter support as a dev-dependency for snapshotting diagnostic publish payloads.

Documentation:
- Document the canonical diagnostic publish path, ownership, and composition rules in the developers guide.

Tests:
- Add snapshot tests pinning the published `PublishDiagnosticsParams` for representative feature and Rust files, normalizing temp directory paths.
- Add property-based tests ensuring publish parameters preserve the computed diagnostics’ count, order, emptiness semantics, and URI target.
- Add unit tests covering behavior when a feature index is missing to ensure publishing is skipped.

### CODE REVIEW

**Comment by buzzybee-df12** [comment]
@coderabbitai review

File: crates/rstest-bdd-server/src/handlers/diagnostics/publish.rs [hunk 1]
```diff
@@ -16,44 +20,100 @@ use super::placeholder::compute_signature_mismatch_diagnostics;
...
+/// performing any I/O towards the client.
+///
+/// Returns `None` (publishing is skipped) when `compute` declines to produce
+/// diagnostics or when `path` cannot be converted to a URI. An empty
+/// diagnostics vector still yields parameters, because publishing an empty
+/// array clears previously published diagnostics.
+///
+/// Separated from [`publish_with`] so tests can pin the published payload
+/// without a client socket.
+fn prepare_publish(
```
[the remark this answers is not in the document]
Comment (pandalump) [reply crates/rstest-bdd-server/src/handlers/diagnostics/publish.rs:77]: Declining this refactor. The higher-order `prepare_publish`/`publish_with` split is the deliberate goal of this PR (issue #510): a single canonical publication boundary owning the client-socket guard, path-to-URI guard, `PublishDiagnosticsParams` construction, notification, and error logging. Collapsing it back into per-`publish_*` helpers would re-duplicate exactly that prologue/epilogue. CodeRabbit's review on the same code asks for *more* coverage of this boundary (now added as an end-to-end `smoke_lsp` test), not its removal, so keeping the seam is the right call.

File: crates/rstest-bdd-server/src/handlers/diagnostics/publish.rs [hunk 2]
```diff
@@ -79,21 +139,128 @@ pub fn publish_all_feature_diagnostics(state: &ServerState) {
...
+
+    const RUST_SOURCE: &str = concat!(
+        "use rstest_bdd_macros::{given, when};\n\n",
+        "#[given(\"an implemented step\")]\n",
+        "fn implemented(count: u32) {}\n\n",
+        "#[when(\"an unused step\")]\n",
+        "fn unused() {}\n",
+    );
+
+    /// Snapshot settings that normalize the temp-dir portion of file URIs.
```
[the remark this answers is not in the document]
Comment (pandalump) [reply crates/rstest-bdd-server/src/handlers/diagnostics/publish.rs:193]: Keeping `normalize`. This repository uses en-GB **Oxford** spelling (`en-GB-oxendict`), which takes `-ize`/`-ization` endings (normalize, organize) while keeping `-yse`/`-our`. The `-ise` form is in fact rejected by the repo's `typos` spelling gate (`make markdownlint`/`make spelling`), so `normalize` is correct here and `normalise` would fail CI.

File: docs/developers-guide.md [hunk 3]
```diff
@@ -756,6 +756,33 @@ feature-gated regression suite in
 `crates/rstest-bdd-harness-gpui/tests/scenario_name_in_logs.rs` apply the
 attribute to every `GpuiHarness::run`-driving test.
 
+## Canonical diagnostic publish path
+
+All LSP diagnostic publishing in `rstest-bdd-server` flows through the
```
[the remark this answers is not in the document]
Comment (pandalump) [reply docs/developers-guide.md]: Done — "LSP" is now expanded to "Language Server Protocol (LSP)" on first use in the developers' guide.

File: crates/rstest-bdd-server/src/handlers/diagnostics/publish.rs [hunk 4]
```diff
@@ -79,21 +139,130 @@ pub fn publish_all_feature_diagnostics(state: &ServerState) {
...
+            let expected = diagnostics.clone();
+            let params = prepare_publish(&state, &path, move |_, _| Some(diagnostics));
+            let params = params.ok_or_else(|| {
+                TestCaseError::fail("valid path must produce publish params")
+            })?;
+            prop_assert_eq!(&params.diagnostics, &expected);
+            #[expect(clippy::expect_used, reason = "fixed absolute path converts")]
+            let expected_uri = Url::from_file_path(&path).expect("uri");
+            prop_assert_eq!(params.uri, expected_uri);
+        }
```
[the remark this answers is not in the document]
Comment (pandalump) [reply crates/rstest-bdd-server/src/handlers/diagnostics/publish.rs]: @coderabbitai The canonical publication boundary is now exercised end-to-end. `crates/rstest-bdd-server/tests/smoke_lsp/main.rs` adds `smoke_feature_diagnostics_cleared_once_step_implemented`, which drives a live server through the public publishers: an unimplemented step emits a non-empty `publishDiagnostics` notification, and implementing the matching Rust step re-publishes an empty diagnostics array (clearing). This asserts real notification emission and empty-vector clearing through `publish_with`, which the `prepare_publish`-only tests could not observe. Please re-verify with codegraph analysis and confirm whether this resolves the comment, or provide a follow-up recommendation.

**Comment by pandalump** [comment]
@coderabbitai review

File: docs/developers-guide.md [hunk 5]
```diff
@@ -844,6 +844,37 @@ feature-gated regression suite in
...
+- **Permitted call-sites:** the public per-file-kind functions
+  (`publish_feature_diagnostics`, `publish_rust_diagnostics`, and any future
+  variant). New diagnostic publishers must delegate to `publish_with` with a
+  compute closure rather than re-implementing the guards or notify call.
+- **Composition rules:** the compute closure returns
+  `Option<Vec<Diagnostic>>` — `None` skips publishing entirely (used when a
+  feature file has no index, preserving previously published diagnostics),
+  while `Some(vec![])` still publishes so stale diagnostics are cleared.
+  `prepare_publish` separates parameter construction from the notify side
+  effect so tests can pin payloads without a client socket.
```
[the remark this answers is not in the document]
Comment (pandalump) [reply docs/developers-guide.md]: Fixed in the latest commit — added the comma before both coordinating `so` clauses ("still publishes, so stale diagnostics are cleared" and "notify side effect, so tests can pin payloads").

File: crates/rstest-bdd-server/tests/smoke_lsp/main.rs [hunk 6]
```diff
@@ -312,3 +312,80 @@ fn smoke_diagnostics_published_for_unimplemented_step(mut server: ServerHandle)
...
+    let feature_path = dir.join("clearing.feature");
+    std::fs::write(
+        &feature_path,
+        concat!(
+            "Feature: clearing\n",
+            "  Scenario: pending\n",
+            "    Given a pending step\n",
+        ),
+    )
+    .expect("write feature");
```
[the remark this answers is not in the document]
Comment (pandalump) [reply crates/rstest-bdd-server/tests/smoke_lsp/main.rs]: This one does not apply here. `smoke_lsp` is one of the integration-test crates deliberately **excluded** from the `no_std_fs_operations` Whitaker lint in `dylint.toml`:

```toml
[no_std_fs_operations]
excluded_crates = [
    # ...
    # Integration test crates for the language server: each stages feature
    # and Rust fixture files in ambient temp directories so the server can
    # index them via the same ambient paths the LSP client would supply.
    "smoke_lsp",
    # ...
]
```

The new fixture writes use `std::fs::write` exactly as the existing `create_test_files` helper in this same file already does — the LSP client stages files on the ambient filesystem, so there is no capability root to thread a `cap_std::fs_utf8::Dir` from. Threading `cap-std` here would diverge from the established pattern the repo has explicitly sanctioned for these crates. Leaving the direct `std::fs` calls as-is, consistent with `dylint.toml` and the surrounding test code.

File: crates/rstest-bdd-server/src/handlers/diagnostics/publish.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (pandalump) [reply crates/rstest-bdd-server/src/handlers/diagnostics/publish.rs]: The canonical boundary is now exercised end-to-end — this finding predates the added coverage. `crates/rstest-bdd-server/tests/smoke_lsp/main.rs::smoke_feature_diagnostics_cleared_once_step_implemented` drives a **live server** through the public publishers (`publish_feature_diagnostics` / `publish_all_feature_diagnostics`, both of which delegate to `publish_with`):

- **Notification emission** — saving a feature with an unimplemented step yields a real `textDocument/publishDiagnostics` notification with a non-empty `diagnostics` array, asserted off the wire.
- **Empty-vector clearing** — implementing the matching Rust step re-publishes the feature with an **empty** `diagnostics` array, asserting the clear-on-empty path.

Both assertions observe the actual `notify` call through the client socket, so a regression that skipped `notify`, used the wrong notification, or bypassed the shared boundary would now fail — which the `prepare_publish`-only unit tests could not catch.

- **Feature-index skip** is covered by the existing unit test `missing_feature_index_skips_publishing`, which asserts `prepare_publish` returns `None` (no params built ⇒ no notify) when the feature index is absent. Asserting the *absence* of a wire notification in the live-server test would require a timeout-based negative assertion, which is flaky; the unit-level check is the deterministic equivalent.

Could you confirm whether this resolves the finding, or point to a specific boundary branch still unexercised?

**Comment by buzzybee-df12** [comment]
@coderabbitai review

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already. (Treat documentation and validation coverage as in scope).

### ❌ Failed checks (2 errors, 1 warning)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | The new tests only call prepare\_publish; a broken or no-op publish\_with would still pass, so the notify boundary and empty-vector clearing remain untested. | Add client-backed tests that call publish\_feature\_diagnostics/publish\_rust\_diagnostics and assert publishDiagnostics emission, empty-vector clearing, and feature-index skip. |
| Unit Architecture | ❌ Error | Tests only pin prepare\_publish; no client-backed test exercises publish\_with, notify, or empty-vector clearing. | Add behavioural tests with an injected client to assert publish\_with emits textDocument/publishDiagnostics, skips on missing feature index, and clears on empty diagnostics. |
| Testing (Unit And Behavioural) | ⚠️ Warning | The new tests only call `prepare_publish`; none exercise `publish_with` or the public publishers, and no assertion checks emitted LSP notifications. | Add client-backed behavioural tests that drive the public save/publish path and assert `textDocument/publishDiagnostics`, empty-vector clearing, and feature-index skip behaviour. |

File: crates/rstest-bdd-server/src/handlers/diagnostics/publish.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (pandalump) [reply crates/rstest-bdd-server/src/handlers/diagnostics/publish.rs]: The remaining client-backed coverage is now implemented (not just the smoke feature test): commit `cc708b1`.

- **Missing-feature-index skip** — `missing_feature_index_emits_no_notification_through_client` in `publish.rs` drives a real `async_lsp` main loop, attaches its `ClientSocket` to `ServerState`, calls `publish_feature_diagnostics` for an unindexed absolute path, and asserts the captured transport holds a `window/logMessage` sentinel but **no** `textDocument/publishDiagnostics` — proving the absent index skips the whole boundary without touching prior client diagnostics. No mock bypasses `ClientSocket::notify`.
- **Rust empty-vector clearing** — `smoke_rust_diagnostics_cleared_once_step_referenced` (now in `tests/smoke_lsp/clearing.rs`) drives the live server: an unused Rust step yields a non-empty `publishDiagnostics`; adding a feature that references it and re-saving the Rust file re-publishes an **empty** array for the same Rust URI (method + URI both asserted).
- Feature emission + feature empty-vector clearing remain covered by `smoke_feature_diagnostics_cleared_once_step_implemented`.

Validated: `make check-fmt`, `make markdownlint`, `make lint` (clippy + Whitaker + file-length), and `make test` (1511 nextest + 56 pytest) all pass. Please re-verify.

**Comment by buzzybee-df12** [comment]
@coderabbitai review

File: crates/rstest-bdd-server/src/handlers/diagnostics/publish.rs [hunk 7]
```diff
@@ -79,21 +139,190 @@ pub fn publish_all_feature_diagnostics(state: &ServerState) {
...
+        let scenario = ScenarioBuilder::new().with_single_file_pair(FEATURE_SOURCE, RUST_SOURCE);
+        let params = prepare_publish(&scenario.state, &scenario.rust_path, |state, path| {
+            Some(compute_rust_file_diagnostics(state, path))
+        });
+        #[expect(clippy::expect_used, reason = "rust files always publish")]
+        let params = params.expect("rust file publishes diagnostics");
+        snapshot_settings(scenario.dir.path()).bind(|| {
+            insta::assert_debug_snapshot!("rust_publish_params", params);
+        });
+    }
```
[the remark this answers is not in the document]
Comment (pandalump) [reply crates/rstest-bdd-server/src/handlers/diagnostics/publish.rs]: Done in `1282d0f`. Added a `publish_scenario` `#[fixture]` returning the concrete `SingleFilePairScenario` from `ScenarioBuilder::new().with_single_file_pair(FEATURE_SOURCE, RUST_SOURCE)`, and converted both payload tests to `#[rstest]` taking it as a parameter. Each test keeps its own diagnostic computation (`compute_feature_file_diagnostics` vs the `compute_rust_file_diagnostics` closure), its own `expect` message, and its existing `snapshot_settings(publish_scenario.dir.path()).bind(|| ...)` wrapper, so the snapshots are unchanged. `make check-fmt`, clippy with `-D warnings`, the publish unit tests, and the `smoke_lsp` suite all pass.

**Comment by pandalump** [comment]
@coderabbitai Re the **User-Facing Documentation** pre-merge warning: actioned in `65be6ce`.

`docs/users-guide.md` now covers both behaviours under **Language server → Diagnostics (on save)**, in the "updated incrementally" subsection:

- **Empty-list clearing** — when a recomputation finds nothing to report, the server publishes an empty diagnostic list, so warnings you have just fixed clear from the Problems panel on the next save rather than lingering until the editor or server restarts.
- **Missing-index skip** — a file the server has not indexed is skipped rather than cleared. A `.feature` file not saved since the server started has no index, so saving an unrelated file leaves its existing diagnostics untouched instead of blanking them; saving the file itself indexes it and refreshes them.

Both are written as observable editor behaviour rather than implementation detail; the ownership and call-site rules stay in the developers' guide. `make check-fmt` and `make markdownlint` pass.

**Comment by pandalump** [comment]
@coderabbitai review

**Comment by buzzybee-df12** [comment]
@coderabbitai review

File: crates/rstest-bdd-server/tests/smoke_lsp/clearing.rs [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+            "  Scenario: uses the step\n",
+            "    Given an orphan step\n",
+        ),
+    )
+    .expect("write feature");
+    did_save(&mut server.stdin, &feature_path);
+
+    // Re-saving the Rust file re-publishes it; the step is now used, so its
+    // computed diagnostics are empty and an empty array clears the stale one.
+    did_save(&mut server.stdin, &rust_path);
```
[the remark this answers is not in the document]
Comment (pandalump) [reply crates/rstest-bdd-server/tests/smoke_lsp/clearing.rs:155]: Verified against the current code, and the interstitial wait is not needed — the ordering it would assert is already guaranteed by the transport. No behavioural change made; `455226e` records the reasoning in the test so the question does not have to be re-derived.

### `didSave` is handled synchronously, in arrival order

`crates/rstest-bdd-server/src/main.rs:123` registers the handler as a synchronous closure:

```rust
.notification::<notification::DidSaveTextDocument>(|st, params| {
    handle_did_save_text_document(st, params);
    ControlFlow::Continue(())
})
```

It takes `&mut ServerState` and returns `ControlFlow` rather than a future, so it runs to completion inline on the main loop. `handle_did_save_text_document` then indexes and publishes inline, with nothing deferred or spawned — `handlers/text_document.rs:55` upserts the index and `:57` publishes immediately:

```rust
state.upsert_feature_index(index);
publish_feature_diagnostics(state, path);
```

Both `didSave` notifications travel the same stdin stream to the same loop, so the server cannot begin reading the Rust re-save before the feature save has been indexed and published. The `--debounce-ms` setting applies to watched-file change processing, not to `didSave` dispatch.

### The sibling test's wait is an assertion, not a barrier

The two tests are already symmetric in structure. Each saves its subject file, asserts a **non-empty** publication for the subject URI, then triggers recomputation and asserts an **empty** one:

| | subject | assert non-empty | trigger | assert empty |
| :-- | :-- | :-- | :-- | :-- |
| feature test | feature | feature URI | save Rust | feature URI |
| rust test | Rust | Rust URI | save feature, re-save Rust | Rust URI |

In the feature test, the wait after the first save exists because a non-empty publication for the subject is one of the things that test is pinning. The Rust test has that same assertion, at lines 122–131. The feature save it makes later is setup for the subject's recomputation, not a subject in its own right — which is why it has no paired assertion. Adding an unasserted wait there would imply a race to a future reader, and its failure mode would be an opaque timeout rather than a named assertion.

The extra re-save the Rust test needs is a real asymmetry in the server, not the test: `handle_feature_file_save` publishes only for the saved feature, whereas `handle_rust_file_save` calls `publish_all_feature_diagnostics` as well. Saving a feature therefore never republishes Rust diagnostics, so the Rust file must be saved again for its own diagnostics to be recomputed.

### Receive budget is not at risk either

`MAX_RECV_MESSAGES` is 20 and `recv_notification_matching` skips non-matching messages. Each test gets a fresh `TempDir`, so exactly one feature exists. Message accounting for the Rust test:

1. save Rust → `publish_all_feature_diagnostics` (no features indexed yet, 0 messages) + `publish_rust_diagnostics` → 1 message, consumed by the matcher.
2. save feature → 1 message for the feature URI, left queued.
3. re-save Rust → 1 message for the feature URI + 1 for the Rust URI.

The final matcher skips two feature-URI messages and matches the Rust one: 3 messages against a budget of 20.

### Empirical check

Both clearing tests, 25 consecutive runs:

```
25 consecutive runs of both clearing tests: pass=25 fail=0
```

Full gates on the branch: `make check-fmt`, `make lint`, `make typecheck`, `make markdownlint`, and `make test` (1513 passed, 7 skipped, 0 failed).

### The general rule

For a synchronous, single-stream LSP notification handler, sequencing two `didSave` notifications is already ordered by the transport; an interstitial receive adds no ordering, only an unasserted wait whose failure mode is a timeout. A wait between client notifications is worth adding when the handler is asynchronous, when work is deferred to a debouncer or spawned task, or when the intermediate publication is itself part of the contract under test. None of those hold here. Structural symmetry between two tests is worth having, but not at the cost of asserting a guarantee the transport supplies — the load-bearing symmetry is in what each test pins about its own subject URI, and that is already present.

[the remark this answers is not in the document]
Comment (leynos) [reply crates/rstest-bdd-server/tests/smoke_lsp/clearing.rs:155]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by buzzybee-df12** [comment]
@coderabbitai review
